File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys

class Solution:
    def __init__(self):
        self.input = sys.stdin.readline
        self.output = sys.stdout.write

    def determination(self, trees, m, h):
        sum = 0
        for x in trees:
            if x > h:
                sum += x - h
        return sum >= m

    def answer(self):
        n, m = map(int, self.input().split())
        trees = [int(i) for i in self.input().split()]

        left, right, ans = 0, 2000000000, 0

        while left <= right:
            mid = (left + right) // 2
            
            if self.determination(trees, m, mid):
                #자를 수 있는 나무 길이 중 최소가 될때까지 실행
                ans = mid
                left = mid + 1
            else:
                right = mid - 1

        self.output(f'{ans}\n')
                
s = Solution()
s.answer()

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] temp = br.readLine().split(" ");
        int n = Integer.parseInt(temp[0]);
        int m = Integer.parseInt(temp[1]);

        String[] inputs = br.readLine().split(" ");
        List<Integer> nums = new ArrayList<>();

        for(int i = 0; i < n ; i++) nums.add(Integer.parseInt(inputs[i]));

        int left = 0;
        int right = 2_000_000_000;
        int mid = 0;
        int answer = 0;
        double temp_res = 0;

        while(left <= right){
            temp_res = 0;
            mid = (left + right) / 2; 

            for(int num : nums) temp_res += num > mid ? (num - mid) : 0;
            
            if(temp_res >= m){
                answer = mid;
                left = mid + 1;
                continue;
            }
            right = mid - 1;
        }
        
        System.out.println(answer);
        br.close();
    }
}